sql/src/main/java/org/apache/druid/sql/calcite/planner/DruidPlanner.java

```diff
@@ -765,13 +786,65 @@ static ParsedNodes create(final SqlNode node) throws ValidationException
       if (query.getKind() == SqlKind.INSERT) {
         insert = (SqlInsert) query;
         query = insert.getSource();
+
+        // Processing to be done when the original query has either of the PARTITION BY or CLUSTER BY clause
+        if (insert instanceof DruidSqlInsert) {
```

Comment:
When will it not be a DruidSqlInsert?

Reply:
Updated with the context in the previous comment. Will make the necessary changes to clean up the if-else mess.

Reply:
On removing the if-else statements, the test case like: `EXPLAIN PLAN FOR INSERT INTO dst SELECT * FROM EXTERN(...)` was failing.
This is because SqlExplain(), still doesn't know about the existence of the DruidSqlInsert node, and returns a SqlInsert() on such queries. Since we don't want the queries to fail on them, it is necessary to keep the unnessecary if-else checks. (I have cleaned up some code in the ftl file though).

Once `EXPLAIN PLAN FOR` works with the modified syntax, there would no longer be any requirements for the check. (I do have a patch that does the same, however, it involves rewriting a small part of the code from the Calcite parser, which I am wary of since we wanna avoid that). I will update the patch on the comments section of PR later (it might have become obsolete due to changing of the keywords).

Reply:
Sounds like there needs to be a parser change so that the production rule above reduces using the Druid-specific node rather than the general one. It would see that `INSERT ...` should reduce to use the Druid node. Then `EXPLAIN` should reference the `INSERT` rule and receive the Druid-specific node. Any idea why that is not happening here?

Reply:
Hi, as mentioned in the caveats section, what this PR achieves is adding a new top-level rule that is druid's explain the syntax. The issue is that explain is another top-level rule, which doesn't reference the druid's explain rule. Therefore `EXPLAIN PLAN FOR <druid's syntax>` is a syntactical error. 

To make it work, we can create a Druid's explain rule like: 
```
+SqlNode DruidSqlExplain() :
+{
+    SqlNode stmt;
+    SqlExplainLevel detailLevel = SqlExplainLevel.EXPPLAN_ATTRIBUTES;
+    SqlExplain.Depth depth;
+    final SqlExplainFormat format;
+}
+{
+    <EXPLAIN> <PLAN>
+    [ detailLevel = ExplainDetailLevel() ]
+    depth = ExplainDepth()
+    (
+        LOOKAHEAD(2)
+        <AS> <XML> { format = SqlExplainFormat.XML; }
+    |
+        <AS> <JSON> { format = SqlExplainFormat.JSON; }
+    |
+        { format = SqlExplainFormat.TEXT; }
+    )
+    <FOR> stmt = SqlQueryOrDmlOrDruidQuery() {
+        return new SqlExplain(getPos(),
+            stmt,
+            detailLevel.symbol(SqlParserPos.ZERO),
+            depth.symbol(SqlParserPos.ZERO),
+            format.symbol(SqlParserPos.ZERO),
+            nDynamicParams);
+    }
+}
+
+
+SqlNode SqlQueryOrDmlOrDruidQuery() :
+{
+    SqlNode stmt;
+}
+{
+    (
+        stmt = DruidSqlInsert()
+    |
+        stmt = SqlQueryOrDml()
+    ) { return stmt; }
+}
```
i.e. the `SqlQueryOrDmlOrDruidQuery` clause is new thing. But this should IMO be for another PR, since this does copy syntax from the `Parser.jj`. (That's why I refrained from adding it in the first place here).

Reply:
Makes sense to me to do it in a follow up. We do need to do it, though, so EXPLAIN on INSERT works properly.